### PR TITLE
Gutenberg: Trash Post

### LINKED ISSUE #27739: Gutenberg: Trash post doesn't work
#### Steps to reproduce
1. Starting at URL: https://wpcalypso.wordpress.com/gutenberg/
2. Start creating new post 
3. In Sidebar select Document -> Status & Visibility
4. Click on Move to trash

#### What I expected
To trash a post.

#### What happened instead
Post and Editor are still present, post is not trashed. 

#### Browser / OS version
Chrome 69 / OS X

#### Screenshot / Video
screen shot 2018-10-10 at 14 13 38

### PR DESCRIPTION
**Pull request by rodrigoi**
Fixes #27739

#### Changes proposed in this Pull Request

* Updates the `wpcomProxyMiddleware` to handle `DELETE` requests.
* Updates the `BrowserURL` component to handle `postStatus` changing to `trash`.

#### Testing instructions

* Create a new page or post with Gutenberg and publish it.
* Click the _Move to trash_ link on the _Document_ > _Status & Visibility_ section.

oct-23-2018 22-11-41

### CODE REVIEW

**Comment by alisterscott** [comment]
Sorry IE11 canaries were busted for a bit - fixed now

File: client/gutenberg/editor/browser-url/index.js [hunk 1]
```diff
@@ -33,6 +35,11 @@ export class BrowserURL extends Component {
 		) {
 			this.props.replaceHistory( `${ currentRoute }/${ postId }` );
 		}
+
+		if ( postStatus === 'trash' && endsWith( currentRoute, `/${ postId }` ) ) {
+			const postType = startsWith( currentRoute, '/gutenberg/post/' ) ? 'posts' : 'pages';
```
Comment (mmtr) [inline client/gutenberg/editor/browser-url/index.js]: You can get the post type from the state and pass it as a prop. The `edit-post` package is doing something similar: https://github.com/WordPress/gutenberg/blob/471e581f0467adaf823dbfabee080ba050a2118f/packages/edit-post/src/components/browser-url/index.js#L96

Comment (Copons) [reply client/gutenberg/editor/browser-url/index.js]: Also please consider the CPT.
Their path format is something like this:
`/types/feedback` for the post list
`/edit/feedback` for the editor (or, of course, `/gutenberg/edit/feedback` for Guten)

Comment (rodrigoi) [reply client/gutenberg/editor/browser-url/index.js]: I kind of ignored other post types. This was more of a quick fix than a something else. I've updated the checks on dff24b9f58a2de80dfed43fbc5ee5da57d514c6d.

**Comment by rodrigoi** [comment]
@mmtr @Copons thanks for the review! Code updated, tested on pages, posts, feedback and testimonials. Also, I replicated the `/my` logic from `post-editor.jsx`

https://github.com/Automattic/wp-calypso/blob/853cf54bfee2fa384bff6b65dd67ee6e982dcd39/client/post-editor/post-editor.jsx#L570-L572

File: client/gutenberg/editor/browser-url/index.js [hunk 2]
```diff
@@ -6,13 +6,26 @@
...
  */
+import { getSelectedSiteId } from 'state/ui/selectors';
+import { getSiteSlug, isJetpackSite, isSingleUserSite } from 'state/sites/selectors';
+
 import getCurrentRoute from 'state/selectors/get-current-route';
-import { replaceHistory } from 'state/ui/actions';
+import { navigate, replaceHistory } from 'state/ui/actions';
+
+const getPostTypeTrashUrl = ( postType, siteSlug, isSiteJetpack, isSiteSingleUser ) => {
+	const postTypeUrl = get( { page: 'pages', post: 'posts' }, postType, `types/${ postType }` );
```
Comment (Copons) [inline client/gutenberg/editor/browser-url/index.js]: Smart! 🤓

File: client/gutenberg/editor/api-middleware/index.js [hunk 3]
```diff
@@ -89,10 +89,17 @@ export const wpcomPathMappingMiddleware = ( options, next, siteSlug ) => {
...
-	if ( method !== 'GET' ) {
-		return wpcom.req.post.bind( wpcom.req );
+	/* wpcom.req.del is just a wrapper around wpcom.req.post
+	 * and it always uses the `POST` method.
+	 * https://github.com/Automattic/wpcom.js/blob/master/lib/util/request.js#L70
+	 *
+	 * Instead we use wpcom.req.get for `DELETE`, that passes the method
+	 * along with the rest of the request parameters.
+	 */
+	if ( includes( [ 'GET', 'DELETE' ], method ) ) {
```
Comment (dmsnell) [inline client/gutenberg/editor/api-middleware/index.js]: two things: we can use `Array.protype.includes` here thanks to our Babel transform

why are we binding `DELETE` to a `GET`?

> that passes the method along with the rest of the request parameters

are you talking about an HTTP envelope? I'm not seeing that anywhere here in the code.

**Pull request merged**